File: number_guessing_game/number_guessing_game.cpp
#include <iostream>
#include <limits>
#include <random>

int main(int argc, char** argv) 
{
    int lower = 1;
    int upper = 100;
    std::random_device rnd;
    std::mt19937 gen(rnd());
    std::uniform_int_distribution<> distrib(lower, upper);
    const int n = distrib(gen);

    int input = 0;
    while (true)
    {
        std::cout << "Make a guess (" << lower << " - " << upper << "): ";
        if (!(std::cin >> input) || input < lower || input > upper)
        {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            continue;
        }
        
        if (input > n) 
        {
            std::cout << input << " is too large" << std::endl;
            upper = input - 1;
        } 
        else if (input < n) 
        {
            std::cout << input << " is too small" << std::endl;
            lower = input + 1;
        } 
        else 
        {
            std::cout << "You are correct" << std::endl;
            break;
        }
    }

    return 0;
}

File: number_guessing_game/number_guessing_game.lua
math.randomseed(os.time())
lower = 1
upper = 100
n = math.random(lower, upper)

input = nil
while true do
    io.write(string.format("Make a guess (%d - %d): ", lower, upper))

    input = tonumber(io.read())
    if input == nil or input < lower or input > upper then
        goto continue
    end

    if input > n then
        print(string.format("%d is too large", input))
        upper = input - 1
    elseif input < n then
        print(string.format("%d is too small", input))
        lower = input + 1
    else
        print("You are correct")
        break
    end

    ::continue::
end